File: frontend/src/types/transferencia.ts
/**
 * Espelha 1:1 `app/schemas/transferencia.py` (backend). Ver
 * docs/analise-arquitetural-transferencias-frontend.md, seções 0 e 4.
 *
 * Sem `TransferenciaUpdate` - decisão do backend, não uma omissão: os
 * campos estruturais (`conta_origem_id`/`conta_destino_id`/`valor`/`data`)
 * são imutáveis após a criação (editar exigiria refazer o cálculo de saldo
 * das duas contas envolvidas). A única transição de estado é a ação
 * `cancelar` (`POST /transferencias/{id}/cancelar`), sem payload.
 */
export interface TransferenciaRead {
  id: number;
  conta_origem_id: number;
  conta_destino_id: number;
  valor: string;
  data: string;
  descricao: string | null;
  ativo: boolean;
}

export interface TransferenciaCreate {
  conta_origem_id: number;
  conta_destino_id: number;
  valor: string;
  data: string;
  descricao?: string | null;
}

/** `GET /transferencias` - parâmetros reais que o backend aceita além de
 * paginação (diferente de Transação, que filtra por período/tipo/status/
 * categoria/conta/cartão de verdade - aqui o filtro server-side é mais
 * enxuto). `conta_id` (genérico, não específico de Meta) foi acrescentado
 * no Refatoramento de Metas/Transferências
 * (docs/analise-arquitetural-metas-transferencias.md, seção 4.2) - usado
 * para montar o histórico de aportes/resgates do "cofrinho" de uma Meta
 * (`MetaResumoCard`), mas serve qualquer conta. */
export interface TransferenciaFiltros {
  apenas_ativas?: boolean;
  conta_id?: number;
  skip?: number;
  limit?: number;
}

File: backend/app/schemas/transferencia.py
"""Schemas de Transferencia: payloads de entrada e saída de
`app/api/routes/transferencia.py`.

Sem `TransferenciaUpdate` - mesma decisão já usada em `Fatura`/`Parcelamento`:
`conta_origem_id`, `conta_destino_id`, `valor` e `data` são estruturais e
imutáveis após a criação (editar qualquer um exigiria refazer o cálculo de
saldo das contas envolvidas); a única transição válida é a ação `cancelar`
(`POST /transferencias/{id}/cancelar`), não uma edição de campo livre.

Validação estrutural (`conta_origem_id` != `conta_destino_id`) não é feita
aqui via `model_validator` - mora inteiramente em `TransferenciaService`,
mesmo raciocínio já documentado em `app/schemas/transacao.py`: evita
duplicar a regra em duas camadas (aqui é ainda mais direto, já que não
existe um `Update` que precisaria repetir a mesma checagem).
"""
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field

from app.schemas.base import OrmBaseModel


class TransferenciaCreate(BaseModel):
    conta_origem_id: int
    conta_destino_id: int
    valor: Decimal = Field(gt=0)
    data: date
    descricao: str | None = Field(default=None, max_length=200)


class TransferenciaRead(OrmBaseModel):
    id: int
    conta_origem_id: int
    conta_destino_id: int
    valor: Decimal
    data: date
    descricao: str | None
    ativo: bool
